#20327: Device queries now faster when including ConfigContexts

jnovinger opened the pull request:
### Fixes: #20327

Move `.distinct()` from main queryset to tag subquery to eliminate performance bottleneck when querying devices with config contexts.

The `.distinct()` call on the main device queryset was causing PostgreSQL to sort all devices before pagination, resulting in 15x slower API responses for large installations (10k+ devices, 100+ config contexts).

Moving `.distinct()` to the tag subquery eliminates duplicates at their source (`GenericForeignKey` tag relationships) while preserving the fix for issues #5314 and #5387 without impacting overall query performance.

This is also likely to have positive impacts in other places that make use of these queryset methods, including:
- /api/virtualization/virtual-machines
- /dcim/devices/
- /virtualization/virtual-machines/

jnovinger committed:
Fixes #20327: Device queries are now faster when including ConfidContexts
Move .distinct() from main queryset to tag subquery to eliminate
performance bottleneck when querying devices with config contexts.

The .distinct() call on the main device queryset was causing PostgreSQL
to sort all devices before pagination, resulting in 15x slower API
responses for large installations (10k+ devices, 100+ config contexts).

Moving .distinct() to the tag subquery eliminates duplicates at their
source (GenericForeignKey tag relationships) while preserving the fix
for issues #5314 and #5387 without impacting overall query performance.

jnovinger committed:
Add performance regression test for config context annotation

The test verifies that:
- Main device queries do not use expensive DISTINCT operations
- Tag subqueries properly use DISTINCT to prevent duplicates from issue #5387

This ensures the optimization from issue #20327 (moving .distinct() from maintaining
query to tag subquery) cannot be accidentally reverted while maintaining the
correctness guarantees for issues #5314 and #5387.

jeremystretch commented on netbox/extras/tests/test_models.py:
```diff
@@ -621,32 +652,35 @@ def test_multiple_tags_return_distinct_objects_with_separate_config_contexts(sel
-        device.local_context_data = ""
-        with self.assertRaises(ValidationError):
-            device.clean()
+        # connection.queries_log.clear()
```
Can we delete this?

jeremystretch commented:
@fabi125 unfortunately you're proposed change deviates from supported patterns, and addresses only the REST API. We should attempt to devise a solution which addresses the inefficiency within the queryset itself.

fabi125 commented:
> @fabi125 unfortunately you're proposed change deviates from supported patterns, and addresses only the REST API. We should attempt to devise a solution which addresses the inefficiency within the queryset itself.

Of course! Fully supportive of fixing this in a proper and supported way instead of my hack :)

jnovinger committed:
Address PR feedback, clean up new regression test

The new regression test now avoids casting the query to a string and
inspecting the string, which was brittle at best.

The new approach asserts directly against `queryset.distinct` for the
main query and then finds the subquery that we expect to have distinct
set and verifies that is in fact the case.

I also realized that the use of `connection.query_log` was problematic,
in that it didn't seem to return any queries as expected. This meant
that the test was actually not making any assertions since none of the
code inside of the for loop over `device_queries` ever ran.
